### PR TITLE
[ALLUXIO-2633] fix copyToLocal command

### PR DESCRIPTION
**Pull request by Dukecat0613**
https://alluxio.atlassian.net/browse/ALLUXIO-2633
bin/alluxio fs copyToLocal /LICENSE .
Failed to rename /Users/aja/alluxio/..WQnkYKkm_copyToLocal_ to destination .

By adding filename to the destination file can solve the problem.

### CODE REVIEW

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 1]
```diff
@@ -566,11 +567,12 @@ private void copyFileToLocal(AlluxioURI srcPath, AlluxioURI dstPath)
         out.write(buf, 0, t);
         t = is.read(buf);
       }
-      if (!tmpDst.renameTo(dstFile)) {
+      File outputFile = new File(dstFile.getAbsolutePath() + "/" + fileName));
```
Comment (maobaolong) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: `dstFile` has been a file already, why append the fileName to the dstFile?

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: dstFile is just the path not the file. If users just set the dstfolder, dstFile would be the path not the file.

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @maobaolong  the rename function can be true only if we set the destination filename, say rename (new File('A.txt'), new File('B.txt')). It will return false  if rename (new File('A.txt'), new File('/Users/Alluxio'))

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 2]
```diff
@@ -552,6 +552,7 @@ private void copyToLocal(AlluxioURI srcPath, AlluxioURI dstPath) throws AlluxioE
   private void copyFileToLocal(AlluxioURI srcPath, AlluxioURI dstPath)
       throws AlluxioException, IOException {
     File dstFile = new File(dstPath.getPath());
+    String fileName = srcPath.getPath().substring(srcPath.getPath().lastIndexOf("/"));
```
Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @maobaolong `srcPath.getName()` works fine.

Comment (maobaolong) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: use srcPath.getName() to get the name

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @maobaolong you are awesome! I did not even notice the function.

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @maobaolong  updated

**Comment by maobaolong** [comment]
- If users execute `alluxio fs copyToLocal /a.txt /home/mbl/a.txt`, what will happend?
`outputFile` maybe `/home/mbl/a.txt/a.txt`

- If users execute `alluxio fs copyToLocal /a.txt /home/mbl`, what will happend?
`tmpDst` maybe `/home/mbl.v1MGmY9u_copyToLocal_ `, and usually a common user have no permission in `/home`

- Do not use `/` to build a file path, use `PathUtils.concatPath` instead.

- java.io.File class have another constructor `public File(String parent, String child)`.

**Comment by Dukecat0613** [comment]
@maobaolong Thanks for your notification. Your information is valuable. I will continue to fix tomorrow

**Comment by maobaolong** [comment]
@Dukecat0613  Thank you for your contribution. Waiting for your next commit. 

BTW, why this PR never invoke any jenkins test job? @yupeng9 @aaudiber @calvinjia

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by Dukecat0613** [comment]
Jenkins, test this please

**Comment by Dukecat0613** [comment]
@maobaolong  

- `dstFile.getAbsolutePath()` will only return the path not including the filename. 

- I place the tmp file in the `HOME` instead.

- Your advice is awesome, replaced with PathUtils.concatPath

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 3]
```diff
@@ -554,7 +555,7 @@ private void copyFileToLocal(AlluxioURI srcPath, AlluxioURI dstPath)
     File dstFile = new File(dstPath.getPath());
     String randomSuffix =
         String.format(".%s_copyToLocal_", RandomStringUtils.randomAlphanumeric(8));
-    File tmpDst = new File(dstFile.getAbsolutePath() + randomSuffix);
+    File tmpDst = new File(Name.HOME + randomSuffix);
```
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: Why did we change this from `dstFile.getAbsolutePath`?

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: if users set the dstPath is `/home/username`
`File tmpDst = new File(dstFile.getAbsolutePath() + randomSuffix);` will do not have the permission to write.

Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: If the user runs `cp file /home/username`, could we make the temporary file `/home/username/file.tmp`, then rename it to `/home/username/file` at the end? That's how the current implementation does things for absolute paths.

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @maobaolong no, it will generate `/home/username.copyToLocal` not `/home/username/copyToLocal`

Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: are you talking about the current behavior of the `master` branch, or the behavior after your change? I'm suggesting that we should have a convention of naming the temp file by taking the destination file (not directory) and appending a random suffix.

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @aaudiber  I am talking about the current master branch. If we want to follow the convention, I can do it. You mean, say if the command is `bin/alluxio fs copyToLocal /LICENSE /home/username`, the tmpfile should be`/home/username.sda3dfg35_`? or `/home/username/.sda3dfg35`

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @aaudiber if you mean `/home/username.sda3dfg35_`, it will cause permission error

Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: assuming `username` is a directory, I mean `/home/username/LICENSE.sda3dfg35_`

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @aaudiber all right, I will try this way right now.

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @aaudiber fixed, it works.

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 4]
```diff
@@ -555,7 +555,8 @@ private void copyFileToLocal(AlluxioURI srcPath, AlluxioURI dstPath)
     File dstFile = new File(dstPath.getPath());
     String randomSuffix =
         String.format(".%s_copyToLocal_", RandomStringUtils.randomAlphanumeric(8));
-    File tmpDst = new File(Name.HOME + randomSuffix);
+    File tmpDst = new File(PathUtils.concatPath( dstFile.getAbsolutePath(),
```
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: does this work if `dstPath` is a file instead of a directory?

**Comment by aaudiber** [comment]
Jenkins, test this please

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 4 above)
Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @aaudiber  what if the file exists? Should we throw a error or overwrite the file?

Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: If the destination exists as a file we should overwrite it, similar to unix cp

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @aaudiber  Do you think it is necessary to write a `getAbsolutePath` method in the `AlluxioURI`? Because the current `getPath` cannot parse relative path correctly. It would be a great pain if we lost the path information whenever parsing the path from user input.

Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: I think we can break fixing `copyToLocal` into two parts:
1. Make it possible to copy a file into a directory by setting the directory as the destination, e.g. `bin/alluxio fs copyToLocal /LICENSE /home/username`.
2. Make it possible to use relative paths, e.g. `bin/alluxio fs copyToLocal /LICENSE dir` or `bin/alluxio fs copyToLocal /LICENSE .`

For this PR I'd like to focus on (1) and leave (2) for a second PR. When we do address (2), I think we can do it by changing `AlluxioURI` parsing so that when the scheme is `file://` it converts relative paths to absolute paths.

Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @aaudiber  

- The latest code I commit can recognize whether the target path is a folder or a existing file or a target filename. 

- The code I wrote is not that so elegant, I think. If we rewrite the `AlluixoURI` parsing, it can be more intuitive and more readable. 

- The tests I used are ```bin/alluxio fs copyToLocal /LICENSE `pwd`/LICENSE```; `bin/alluxio fs copyToLocal /LICENSE LICENSE`; `bin/alluxio fs copyToLocal /LICENSE .` ; `bin/alluxio fs copyToLocal /LICENSE testfilename`; But will fail `bin/alluxio fs copyToLocal /LICENSE ./LICENSE`(the current parsing connot recongnize `./`)

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 5]
```diff
@@ -554,7 +554,34 @@ private void copyFileToLocal(AlluxioURI srcPath, AlluxioURI dstPath)
...
+      if (dstPath.getAuthority().equals(".")) {
+        fileName = srcPath.getName();
+      } else {
+        fileName = dstPath.getAuthority();
+      }
+    } else {
+      fileName = dstPath.getName();
+    }
+
+    if (new File(dstAbsolute).exists()) {
```
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: when the destination is a file, we can treat it the same as if it doesn't exist, so we only need two cases based on `isFile()`

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 6]
```diff
@@ -554,7 +554,34 @@ private void copyFileToLocal(AlluxioURI srcPath, AlluxioURI dstPath)
     File dstFile = new File(dstPath.getPath());
     String randomSuffix =
         String.format(".%s_copyToLocal_", RandomStringUtils.randomAlphanumeric(8));
-    File tmpDst = new File(dstFile.getAbsolutePath() + randomSuffix);
+    File tmpDst;
+    File outputFile;
+    String dstAbsolute = dstFile.getAbsolutePath();
+    String fileName;
+    if (dstPath.getPath().equals("")) {
```
Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @aaudiber  I think it is necessary to handle relative path. Bcz the `copyFromLocal` command also does not recognize relative path

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 5 above)
Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @aaudiber yes, you are right. Fixing

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 6 above)
Comment (Dukecat0613) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: @aaudiber  I am going to try to add method in `AlluxioURI`

**Comment by Dukecat0613** [comment]
@aaudiber update for `AlluxioURI.getAbsolutePath()`

File: core/common/src/main/java/alluxio/AlluxioURI.java [hunk 7]
```diff
@@ -448,4 +449,18 @@ public String toString() {
...
+
+  /**
+   * Gets the path component of the {@link AlluxioURI}.
+   *
+   * @return the absolute path
+   */
+  public String getAbsolutePath() {
+    if (mUri.getAuthority() != null) {
+      return PathUtils.concatPath(System.getProperty("user.dir"),
+              mUri.getAuthority().equals(".") ? mUri.getPath() : mUri.getAuthority());
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/AlluxioURI.java]: `.` is technically a part of the URI path, not the authority, so this is relying on a bug in `AlluxioURI` where it gets confused by `.` and puts it in the authority. We should either fix `AlluxioURI` to completely support relative paths (preferably in a separate PR), or not support relative paths at all.

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 8]
```diff
@@ -551,10 +551,19 @@ private void copyToLocal(AlluxioURI srcPath, AlluxioURI dstPath) throws AlluxioE
...
       throws AlluxioException, IOException {
-    File dstFile = new File(dstPath.getPath());
+    File dstFile = new File(dstPath.getAbsolutePath());
     String randomSuffix =
         String.format(".%s_copyToLocal_", RandomStringUtils.randomAlphanumeric(8));
-    File tmpDst = new File(dstFile.getAbsolutePath() + randomSuffix);
+    File tmpDst;
+    File outputFile;
+    if (dstFile.isDirectory()) {
+      tmpDst = new File(PathUtils.concatPath(dstFile.getAbsolutePath(),
```
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: we can simplify this a little by using if-else to determine `outputFile`, then set `tmpDst` to be `outputFile` plus a random suffix outside of the if-else

File: core/common/src/main/java/alluxio/AlluxioURI.java (hunk 7 above)
Comment (Dukecat0613) [inline core/common/src/main/java/alluxio/AlluxioURI.java]: @aaudiber  `AlluxioURI` is fully based on the `alluxio.URI`. You mean, if we want to fix the `AlluxioURI`, we should completely rewrite the how `alluxio.URI` parse the  path?

Comment (Dukecat0613) [inline core/common/src/main/java/alluxio/AlluxioURI.java]: @aaudiber  If so, I think I should close the PR.

Comment (aaudiber) [inline core/common/src/main/java/alluxio/AlluxioURI.java]: I think this PR is still valuable for `copyToLocal` work with absolute-path directories work. You are right that we need to change the `URI.Factory` parsing methods to be able to resolve relative paths with the `file://` scheme into absolute paths. I think that's the simplest path forward for supporting all relative paths, which is really important for making the `cp` command easy to use

Comment (Dukecat0613) [inline core/common/src/main/java/alluxio/AlluxioURI.java]: @aaudiber  Ok, So do you think it is difficult to change the method to parse `file://` path? I am very interested in this issue.

Comment (aaudiber) [inline core/common/src/main/java/alluxio/AlluxioURI.java]: You'll need to be careful and write some good unit tests to make sure the parsing doesn't miss any edge cases, but it shouldn't be too difficult. I've created a JIRA ticket to track the issue, if you can start working on it please assign it to yourself https://alluxio.atlassian.net/browse/ALLUXIO-2648.

Lets handle it in a separate PR so that it doesn't block the copy-to-directory fix.

**Comment by Dukecat0613** [comment]
@aaudiber  do you have any idea about how the `AlluxioURI` needs to be changed? The `URI.Factory.create` is almost the same as `HDFS path parse` part. So do you have some thoughts or some hints you have?

**Comment by aaudiber** [comment]
@Dukecat0613 lets move the discussion to the relevant JIRA ticket and make the goal of this PR just to address https://alluxio.atlassian.net/browse/ALLUXIO-2633, not to deal with relative paths. I've posted an idea there
